File: mobile/src/main/scala/com/github/shadowsocks/ProfileConfigFragment.scala
/*******************************************************************************/
/*                                                                             */
/*  Copyright (C) 2017 by Max Lv <max.c.lv@gmail.com>                          */
/*  Copyright (C) 2017 by Mygod Studio <contact-shadowsocks-android@mygod.be>  */
/*                                                                             */
/*  This program is free software: you can redistribute it and/or modify       */
/*  it under the terms of the GNU General Public License as published by       */
/*  the Free Software Foundation, either version 3 of the License, or          */
/*  (at your option) any later version.                                        */
/*                                                                             */
/*  This program is distributed in the hope that it will be useful,            */
/*  but WITHOUT ANY WARRANTY; without even the implied warranty of             */
/*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the              */
/*  GNU General Public License for more details.                               */
/*                                                                             */
/*  You should have received a copy of the GNU General Public License          */
/*  along with this program. If not, see <http://www.gnu.org/licenses/>.       */
/*                                                                             */
/*******************************************************************************/

package com.github.shadowsocks

import android.app.Activity
import android.content._
import android.os.{Build, Bundle, UserManager}
import android.support.design.widget.Snackbar
import android.support.v14.preference.SwitchPreference
import android.support.v7.app.AlertDialog
import android.support.v7.preference.{Preference, PreferenceDataStore}
import android.support.v7.widget.Toolbar.OnMenuItemClickListener
import android.text.TextUtils
import android.view.MenuItem
import com.github.shadowsocks.ShadowsocksApplication.app
import com.github.shadowsocks.database.Profile
import com.github.shadowsocks.plugin._
import com.github.shadowsocks.preference.{IconListPreference, OnPreferenceDataStoreChangeListener, PluginConfigurationDialogFragment}
import com.github.shadowsocks.utils.{Action, Key}
import com.takisoft.fix.support.v7.preference.{EditTextPreference, PreferenceFragmentCompatDividers}

object ProfileConfigFragment {
  private final val REQUEST_CODE_PLUGIN_CONFIGURE = 1
  private final val FRAGMENT_DIALOG_TAG = "android.support.v7.preference.PreferenceFragment.DIALOG"
}

class ProfileConfigFragment extends PreferenceFragmentCompatDividers with OnMenuItemClickListener
  with OnPreferenceDataStoreChangeListener {
  import ProfileConfigFragment._

  private var profile: Profile = _
  private var isProxyApps: SwitchPreference = _
  private var plugin: IconListPreference = _
  private var pluginConfigure: EditTextPreference = _
  private var pluginConfiguration: PluginConfiguration = _

  override def onCreatePreferencesFix(bundle: Bundle, key: String) {
    getPreferenceManager.setPreferenceDataStore(app.dataStore)
    app.profileManager.getProfile(getActivity.getIntent.getIntExtra(Action.EXTRA_PROFILE_ID, -1)) match {
      case Some(p) =>
        profile = p
        profile.serialize(app.dataStore)
      case None => getActivity.finish()
    }
    addPreferencesFromResource(R.xml.pref_profile)
    if (Build.VERSION.SDK_INT >= 25 && getActivity.getSystemService(classOf[UserManager]).isDemoUser) {
      findPreference(Key.host).setSummary("shadowsocks.example.org")
      findPreference(Key.remotePort).setSummary("1337")
      findPreference(Key.password).setSummary("\u2022" * 32)
    }
    isProxyApps = findPreference(Key.proxyApps).asInstanceOf[SwitchPreference]
    if (Build.VERSION.SDK_INT < 21) isProxyApps.getParent.removePreference(isProxyApps) else {
      isProxyApps.setEnabled(app.usingVpnMode)
      isProxyApps.setOnPreferenceClickListener(_ => {
        startActivity(new Intent(getActivity, classOf[AppManager]))
        isProxyApps.setChecked(true)
        false
      })
    }
    plugin = findPreference(Key.plugin).asInstanceOf[IconListPreference]
    pluginConfigure = findPreference(Key.pluginConfigure).asInstanceOf[EditTextPreference]
    plugin.unknownValueSummary = getString(R.string.plugin_unknown)
    plugin.setOnPreferenceChangeListener((_, value) => {
      val selected = value.asInstanceOf[String]
      pluginConfiguration = new PluginConfiguration(pluginConfiguration.pluginsOptions, selected)
      app.dataStore.plugin = pluginConfiguration.toString
      app.dataStore.dirty = true
      pluginConfigure.setEnabled(!TextUtils.isEmpty(selected))
      pluginConfigure.setText(pluginConfiguration.selectedOptions.toString)
      if (!PluginManager.fetchPlugins()(selected).trusted)
        Snackbar.make(getView, R.string.plugin_untrusted, Snackbar.LENGTH_LONG).show()
      true
    })
    pluginConfigure.setOnPreferenceChangeListener(onPluginConfigureChanged)
    initPlugins()
    app.listenForPackageChanges(initPlugins())
    app.dataStore.registerChangeListener(this)
  }

  def initPlugins() {
    val plugins = PluginManager.fetchPlugins()
    plugin.setEntries(plugins.map(_._2.label).toArray)
    plugin.setEntryValues(plugins.map(_._2.id.asInstanceOf[CharSequence]).toArray)
    plugin.setEntryIcons(plugins.map(_._2.icon).toArray)
    plugin.entryPackageNames = plugins.map(_._2.packageName).toArray
    pluginConfiguration = new PluginConfiguration(app.dataStore.plugin)
    plugin.setValue(pluginConfiguration.selected)
    plugin.init()
    plugin.checkSummary()
    pluginConfigure.setEnabled(!TextUtils.isEmpty(pluginConfiguration.selected))
    pluginConfigure.setText(pluginConfiguration.selectedOptions.toString)
  }

  private def onPluginConfigureChanged(p: Preference, value: Any) = try {
    val selected = pluginConfiguration.selected
    pluginConfiguration = new PluginConfiguration(pluginConfiguration.pluginsOptions +
      (pluginConfiguration.selected -> new PluginOptions(selected, value.asInstanceOf[String])), selected)
    app.dataStore.plugin = pluginConfiguration.toString
    app.dataStore.dirty = true
    true
  } catch {
    case exc: IllegalArgumentException =>
      Snackbar.make(getView, exc.getLocalizedMessage, Snackbar.LENGTH_LONG).show()
      false
  }

  def onPreferenceDataStoreChanged(store: PreferenceDataStore, key: String): Unit =
    if (key != Key.proxyApps && findPreference(key) != null) app.dataStore.dirty = true

  override def onDestroy() {
    app.dataStore.unregisterChangeListener(this)
    super.onDestroy()
  }

  override def onResume() {
    super.onResume()
    isProxyApps.setChecked(app.dataStore.proxyApps) // fetch proxyApps updated by AppManager
  }

  private def showPluginEditor() {
    val bundle = new Bundle()
    bundle.putString("key", Key.pluginConfigure)
    bundle.putString(PluginConfigurationDialogFragment.PLUGIN_ID_FRAGMENT_TAG, pluginConfiguration.selected)
    val fragment = new PluginConfigurationDialogFragment
    fragment.setArguments(bundle)
    fragment.setTargetFragment(this, 0)
    fragment.show(getFragmentManager, FRAGMENT_DIALOG_TAG)
  }

  override def onDisplayPreferenceDialog(preference: Preference): Unit = if (preference.getKey == Key.pluginConfigure) {
    val intent = PluginManager.buildIntent(pluginConfiguration.selected, PluginContract.ACTION_CONFIGURE)
    if (intent.resolveActivity(getActivity.getPackageManager) != null)
      startActivityForResult(intent.putExtra(PluginContract.EXTRA_OPTIONS,
        pluginConfiguration.selectedOptions.toString), REQUEST_CODE_PLUGIN_CONFIGURE) else {
      showPluginEditor()
    }
  } else super.onDisplayPreferenceDialog(preference)

  override def onActivityResult(requestCode: Int, resultCode: Int, data: Intent): Unit = requestCode match {
    case REQUEST_CODE_PLUGIN_CONFIGURE => resultCode match {
      case Activity.RESULT_OK =>
        val options = data.getStringExtra(PluginContract.EXTRA_OPTIONS)
        pluginConfigure.setText(options)
        onPluginConfigureChanged(null, options)
      case PluginContract.RESULT_FALLBACK => showPluginEditor()
      case _ =>
    }
    case _ => super.onActivityResult(requestCode, resultCode, data)
  }

  override def onMenuItemClick(item: MenuItem): Boolean = item.getItemId match {
    case R.id.action_delete =>
      new AlertDialog.Builder(getActivity)
        .setTitle(R.string.delete_confirm_prompt)
        .setPositiveButton(R.string.yes, ((_, _) => {
          app.profileManager.delProfile(profile.id)
          getActivity.finish()
        }): DialogInterface.OnClickListener)
        .setNegativeButton(R.string.no, null)
        .create()
        .show()
      true
    case R.id.action_apply =>
      saveAndExit()
      true
    case _ => false
  }

  def saveAndExit() {
    profile.deserialize(app.dataStore)
    app.profileManager.updateProfile(profile)
    if (ProfilesFragment.instance != null) ProfilesFragment.instance.profilesAdapter.deepRefreshId(profile.id)
    getActivity.finish()
  }
}

File: mobile/src/main/java/com/github/shadowsocks/ProfileConfigFragment.kt
/*******************************************************************************
 *                                                                             *
 *  Copyright (C) 2017 by Max Lv <max.c.lv@gmail.com>                          *
 *  Copyright (C) 2017 by Mygod Studio <contact-shadowsocks-android@mygod.be>  *
 *                                                                             *
 *  This program is free software: you can redistribute it and/or modify       *
 *  it under the terms of the GNU General Public License as published by       *
 *  the Free Software Foundation, either version 3 of the License, or          *
 *  (at your option) any later version.                                        *
 *                                                                             *
 *  This program is distributed in the hope that it will be useful,            *
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of             *
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the              *
 *  GNU General Public License for more details.                               *
 *                                                                             *
 *  You should have received a copy of the GNU General Public License          *
 *  along with this program. If not, see <http://www.gnu.org/licenses/>.       *
 *                                                                             *
 *******************************************************************************/

package com.github.shadowsocks

import android.app.Activity
import android.content.BroadcastReceiver
import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.os.Parcelable
import android.view.MenuItem
import android.view.View
import androidx.appcompat.app.AlertDialog
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.whenCreated
import androidx.preference.*
import com.github.shadowsocks.database.Profile
import com.github.shadowsocks.database.ProfileManager
import com.github.shadowsocks.plugin.*
import com.github.shadowsocks.preference.*
import com.github.shadowsocks.utils.*
import com.github.shadowsocks.widget.ListListener
import com.google.android.material.snackbar.Snackbar
import kotlinx.android.parcel.Parcelize
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class ProfileConfigFragment : PreferenceFragmentCompat(),
        Preference.OnPreferenceChangeListener, OnPreferenceDataStoreChangeListener {
    companion object PasswordSummaryProvider : Preference.SummaryProvider<EditTextPreference> {
        override fun provideSummary(preference: EditTextPreference?) = "\u2022".repeat(preference?.text?.length ?: 0)

        private const val REQUEST_CODE_PLUGIN_CONFIGURE = 1
        const val REQUEST_UNSAVED_CHANGES = 2
        private const val REQUEST_PICK_PLUGIN = 3
    }

    @Parcelize
    data class ProfileIdArg(val profileId: Long) : Parcelable
    class DeleteConfirmationDialogFragment : AlertDialogFragment<ProfileIdArg, Empty>() {
        override fun AlertDialog.Builder.prepare(listener: DialogInterface.OnClickListener) {
            setTitle(R.string.delete_confirm_prompt)
            setPositiveButton(R.string.yes) { _, _ ->
                ProfileManager.delProfile(arg.profileId)
                requireActivity().finish()
            }
            setNegativeButton(R.string.no, null)
        }
    }

    private var profileId = -1L
    private lateinit var isProxyApps: SwitchPreference
    private lateinit var plugin: PluginPreference
    private lateinit var pluginConfigure: EditTextPreference
    private lateinit var pluginConfiguration: PluginConfiguration
    private lateinit var receiver: BroadcastReceiver
    private lateinit var udpFallback: Preference

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        preferenceManager.preferenceDataStore = DataStore.privateStore
        val activity = requireActivity()
        profileId = activity.intent.getLongExtra(Action.EXTRA_PROFILE_ID, -1L)
        if (profileId != -1L && profileId != DataStore.editingId) {
            activity.finish()
            return
        }
        addPreferencesFromResource(R.xml.pref_profile)
        findPreference<EditTextPreference>(Key.remotePort)!!.setOnBindEditTextListener(EditTextPreferenceModifiers.Port)
        findPreference<EditTextPreference>(Key.password)!!.summaryProvider = PasswordSummaryProvider
        val serviceMode = DataStore.serviceMode
        findPreference<Preference>(Key.remoteDns)!!.isEnabled = serviceMode != Key.modeProxy
        findPreference<Preference>(Key.ipv6)!!.isEnabled = serviceMode == Key.modeVpn
        isProxyApps = findPreference(Key.proxyApps)!!
        isProxyApps.isEnabled = serviceMode == Key.modeVpn
        isProxyApps.setOnPreferenceChangeListener { _, newValue ->
            startActivity(Intent(activity, AppManager::class.java))
            if (newValue as Boolean) DataStore.dirty = true
            newValue
        }
        findPreference<Preference>(Key.metered)!!.apply {
            if (Build.VERSION.SDK_INT >= 28) isEnabled = serviceMode == Key.modeVpn else remove()
        }
        findPreference<Preference>(Key.udpdns)!!.isEnabled = serviceMode != Key.modeProxy
        plugin = findPreference(Key.plugin)!!
        pluginConfigure = findPreference(Key.pluginConfigure)!!
        pluginConfigure.setOnBindEditTextListener(EditTextPreferenceModifiers.Monospace)
        pluginConfigure.onPreferenceChangeListener = this
        pluginConfiguration = PluginConfiguration(DataStore.plugin)
        initPlugins()
        udpFallback = findPreference(Key.udpFallback)!!
        DataStore.privateStore.registerChangeListener(this)

        val profile = ProfileManager.getProfile(profileId) ?: Profile()
        if (profile.subscription == Profile.SubscriptionStatus.Active) {
            findPreference<Preference>(Key.name)!!.isEnabled = false
            findPreference<Preference>(Key.host)!!.isEnabled = false
            findPreference<Preference>(Key.password)!!.isEnabled = false
            findPreference<Preference>(Key.method)!!.isEnabled = false
            findPreference<Preference>(Key.remotePort)!!.isEnabled = false
            plugin.isEnabled = false
            pluginConfigure.isEnabled = false
            udpFallback.isEnabled = false
        }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        listView.setOnApplyWindowInsetsListener(ListListener)
    }

    private fun initPlugins() {
        plugin.value = pluginConfiguration.selected
        plugin.init()
        pluginConfigure.isEnabled = plugin.selectedEntry?.let { it is NoPlugin } == false
        pluginConfigure.text = pluginConfiguration.getOptions().toString()
    }

    private fun showPluginEditor() {
        PluginConfigurationDialogFragment().apply {
            setArg(Key.pluginConfigure, pluginConfiguration.selected)
            setTargetFragment(this@ProfileConfigFragment, 0)
        }.showAllowingStateLoss(parentFragmentManager, Key.pluginConfigure)
    }

    private fun saveAndExit() {
        val profile = ProfileManager.getProfile(profileId) ?: Profile()
        profile.id = profileId
        profile.deserialize()
        ProfileManager.updateProfile(profile)
        ProfilesFragment.instance?.profilesAdapter?.deepRefreshId(profileId)
        if (profileId in Core.activeProfileIds && DataStore.directBootAware) DirectBoot.update()
        requireActivity().finish()
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        receiver = context.listenForPackageChanges(false) {
            lifecycleScope.launch(Dispatchers.Main) {   // wait until changes were flushed
                whenCreated { initPlugins() }
            }
        }
    }

    override fun onResume() {
        super.onResume()
        isProxyApps.isChecked = DataStore.proxyApps // fetch proxyApps updated by AppManager
        val fallbackProfile = DataStore.udpFallback?.let { ProfileManager.getProfile(it) }
        if (fallbackProfile == null) udpFallback.setSummary(R.string.plugin_disabled)
        else udpFallback.summary = fallbackProfile.formattedName
    }

    override fun onPreferenceChange(preference: Preference?, newValue: Any?): Boolean = try {
        val selected = pluginConfiguration.selected
        pluginConfiguration = PluginConfiguration(pluginConfiguration.pluginsOptions +
                (pluginConfiguration.selected to PluginOptions(selected, newValue as? String?)), selected)
        DataStore.plugin = pluginConfiguration.toString()
        DataStore.dirty = true
        true
    } catch (exc: RuntimeException) {
        Snackbar.make(view!!, exc.readableMessage, Snackbar.LENGTH_LONG).show()
        false
    }

    override fun onPreferenceDataStoreChanged(store: PreferenceDataStore, key: String) {
        if (key != Key.proxyApps && findPreference<Preference>(key) != null) DataStore.dirty = true
    }

    override fun onDisplayPreferenceDialog(preference: Preference) {
        when (preference.key) {
            Key.plugin -> PluginPreferenceDialogFragment().apply {
                setArg(Key.plugin)
                setTargetFragment(this@ProfileConfigFragment, REQUEST_PICK_PLUGIN)
            }.showAllowingStateLoss(parentFragmentManager, Key.plugin)
            Key.pluginConfigure -> {
                val intent = PluginManager.buildIntent(plugin.selectedEntry!!.id, PluginContract.ACTION_CONFIGURE)
                if (intent.resolveActivity(requireContext().packageManager) == null) showPluginEditor() else {
                    startActivityForResult(intent
                            .putExtra(PluginContract.EXTRA_OPTIONS, pluginConfiguration.getOptions().toString()),
                            REQUEST_CODE_PLUGIN_CONFIGURE)
                }
            }
            else -> super.onDisplayPreferenceDialog(preference)
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        when (requestCode) {
            REQUEST_PICK_PLUGIN -> if (resultCode == Activity.RESULT_OK) {
                val selected = plugin.plugins.lookup.getValue(
                        data?.getStringExtra(PluginPreferenceDialogFragment.KEY_SELECTED_ID)!!)
                val override = pluginConfiguration.pluginsOptions.keys.firstOrNull {
                    plugin.plugins.lookup[it] == selected
                }
                pluginConfiguration = PluginConfiguration(pluginConfiguration.pluginsOptions, override ?: selected.id)
                DataStore.plugin = pluginConfiguration.toString()
                DataStore.dirty = true
                plugin.value = pluginConfiguration.selected
                pluginConfigure.isEnabled = selected !is NoPlugin
                pluginConfigure.text = pluginConfiguration.getOptions().toString()
                if (!selected.trusted) Snackbar.make(view!!, R.string.plugin_untrusted, Snackbar.LENGTH_LONG).show()
            }
            REQUEST_CODE_PLUGIN_CONFIGURE -> when (resultCode) {
                Activity.RESULT_OK -> {
                    val options = data?.getStringExtra(PluginContract.EXTRA_OPTIONS)
                    pluginConfigure.text = options
                    onPreferenceChange(null, options)
                }
                PluginContract.RESULT_FALLBACK -> showPluginEditor()
            }
            REQUEST_UNSAVED_CHANGES -> when (resultCode) {
                DialogInterface.BUTTON_POSITIVE -> saveAndExit()
                DialogInterface.BUTTON_NEGATIVE -> requireActivity().finish()
            }
            else -> super.onActivityResult(requestCode, resultCode, data)
        }
    }

    override fun onOptionsItemSelected(item: MenuItem) = when (item.itemId) {
        R.id.action_delete -> {
            DeleteConfirmationDialogFragment().withArg(ProfileIdArg(profileId)).show(this)
            true
        }
        R.id.action_apply -> {
            saveAndExit()
            true
        }
        else -> false
    }

    override fun onDetach() {
        requireContext().unregisterReceiver(receiver)
        super.onDetach()
    }

    override fun onDestroy() {
        DataStore.privateStore.unregisterChangeListener(this)
        super.onDestroy()
    }
}
